### ISSUE TITLE
Cleanup: remove `email` field on `User` in Rust code

### ISSUE THREAD

**Issue by carols10cents** [rust-lang/crates.io#1888]
This is for someone who likes fixing compiler errors and knows or wants to learn about diesel :)

As of [this PR](https://github.com/rust-lang/crates.io/pull/1045), I'm *pretty* sure we don't use the `email` field on `User` anywhere directly anymore (and if we aren't, we shouldn't be). Instead, we have an `emails` table that joins to `users`.

So we can clean this field up and be less confusing about where the user's email is.

The first step is removing all usage of this field from the Rust code, but leaving the column on the `users` table in the database. Once the Rust code is no longer using the column, we can safely migrate the database to drop that column.

But this issue is for just the first step. The way I'd go about it, which isn't the only way so feel free to go about this the way you'd like, is to remove [this line](https://github.com/rust-lang/crates.io/blob/0a6d3fb4aa75fbe504cb0c5a8ce1d853ddd244eb/src/models/user.rs#L16) from the `User` struct definition and try to compile.

A lot of the error messages are going to be because the default behavior of diesel is to select all columns of a table and expect the struct definition to match. Removing the email field from the struct will break that assumption. We need to change the Rust code to never select the email field from the database. We have this problem with krate too, which has a column having to do with full text search that we never want to select in the Rust code. 

The way the krate code works is it defines a type alias and a constant named `ALL_COLUMNS`: https://github.com/rust-lang/crates.io/blob/0a6d3fb4aa75fbe504cb0c5a8ce1d853ddd244eb/src/models/krate.rs#L50-L76

And then everywhere that we select a krate from the database, use that constant instead. For example, selecting all crates happens with the `all` function: https://github.com/rust-lang/crates.io/blob/0a6d3fb4aa75fbe504cb0c5a8ce1d853ddd244eb/src/models/krate.rs#L245-L247 which uses the `ALL_COLUMNS` constant and another type alias named `All`: https://github.com/rust-lang/crates.io/blob/0a6d3fb4aa75fbe504cb0c5a8ce1d853ddd244eb/src/models/krate.rs#L81

Other convenience functions like `Crate::by_name`: https://github.com/rust-lang/crates.io/blob/0a6d3fb4aa75fbe504cb0c5a8ce1d853ddd244eb/src/models/krate.rs#L237-L239 chain `filter` calls off `all` to avoid having to duplicate the `.select(ALL_COLUMNS)` bit.

Extracting some helper functions might be an intermediate step that can be done in a separate commit from, but informed by, the fixes to the compilation errors. In other words, the errors can tell you where we're currently selecting users from the database directly that we could maybe be using a helper function instead. Extracting those could be done in a commit that passes tests as it would be a smaller refactoring that wouldn't change behavior. Then the commit that removes the field from the struct would need fewer changes in fewer places, hopefully. This may not actually be the case, or may not actually make the changes easier, and it may be just the way I work. If this paragraph doesn't make sense, ignore it!

**Comment by hbina** [comment]
I would like to work on this!

**Comment by carols10cents** [comment]
@hbina great!! Please let me know if you have any questions!!!

**Comment by hbina** [comment]
So if I am reading your suggestions correctly, the idea is to:

1. take note of where this struct is used by diesel,
2. see if any of these usages are using the email variable,
3. if so, try to refactor out this usage by using a helper function that use some other ways to obtain user's email (?)
4. this will hopefully pave way for future plan to drop email column from the table

Correct?

**Comment by carols10cents** [comment]
>     4\. this will hopefully pave way for future plan to drop `email` column from the table
> 
> 
> Correct?

You've got it!

> This part looks like an authentication thingy. I don't quite understand how it all works but it does not seem to rely on the `email` variable at all? The only variable that matters here seems to be the `id`.

Correct! The error message is complaining about what the database is returning, that the database results don't match what diesel was expecting for the `User` struct. This spot doesn't make use of `user.email` at all. I expect most of the errors to be like this.

**Comment by hbina** [comment]
So, I think the only non-trivial cases are error number 9, 15, and 17. I will copy them here for ease.

9.

https://github.com/rust-lang/crates.io/blob/21e04f3028624485621d577359008560728a37d3/src/controllers/user/session.rs#L118-L141

```
error[E0277]: the trait bound `(i32, std::string::String, std::string::String, std::option::Option<std::string::String>, std::option::Option<std::string::String>, i32): diesel::Queryable<(diesel::sql_types::Integer, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Integer), diesel::pg::Pg>` is not satisfied
   --> src/controllers/user/session.rs:134:22
    |
134 |                     .first(conn)
    |                      ^^^^^ the trait `diesel::Queryable<(diesel::sql_types::Integer, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Integer), diesel::pg::Pg>` is not implemented for `(i32, std::string::String, std::string::String, std::option::Option<std::string::String>, std::option::Option<std::string::String>, i32)`
    |
    = help: the following implementations were found:
              <(A, B, C, D, E, F) as diesel::Queryable<(SA, SB, SC, SD, SE, SF), __DB>>
              <(A, B, C, D, E, F) as diesel::Queryable<diesel::sql_types::Record<(SA, SB, SC, SD, SE, SF)>, diesel::pg::Pg>>
    = note: required because of the requirements on the impl of `diesel::Queryable<(diesel::sql_types::Integer, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Integer), diesel::pg::Pg>` for `models::user::User`
    = note: required because of the requirements on the impl of `diesel::query_dsl::LoadQuery<diesel::PgConnection, models::user::User>` for `diesel::query_builder::SelectStatement<schema::users::table, diesel::query_builder::select_clause::DefaultSelectClause, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::WhereClause<diesel::expression::operators::Eq<schema::users::columns::gh_id, diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>>>, diesel::query_builder::order_clause::NoOrderClause, diesel::query_builder::limit_clause::LimitClause<diesel::expression::bound::Bound<diesel::sql_types::BigInt, i64>>>`
```

This one is interesting. I think what this function does is create a crates.io account  from a Github account. The issue here is that Github struct contains an email variable. I believe the changes that we need to make here is to insert this data to the appropriate table as a seperate function call.

15.

https://github.com/rust-lang/crates.io/blob/21e04f3028624485621d577359008560728a37d3/src/models/user.rs#L84-L100

```
error[E0609]: no field `email` on type `models::user::User`
  --> src/models/user.rs:84:44
   |
84 |             if let Some(user_email) = user.email.as_ref() {
   |                                            ^^^^^ unknown field
   |
   = note: available fields are: `id`, `gh_access_token`, `gh_login`, `name`, `gh_avatar`, `gh_id`

error[E0277]: the trait bound `(i32, std::string::String, std::string::String, std::option::Option<std::string::String>, std::option::Option<std::string::String>, i32): diesel::Queryable<(diesel::sql_types::Integer, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Integer), diesel::pg::Pg>` is not satisfied
   --> src/models/user.rs:128:37
    |
128 |         users::table.find(user_id_).first(conn)
    |                                     ^^^^^ the trait `diesel::Queryable<(diesel::sql_types::Integer, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Integer), diesel::pg::Pg>` is not implemented for `(i32, std::string::String, std::string::String, std::option::Option<std::string::String>, std::option::Option<std::string::String>, i32)`
    |
    = help: the following implementations were found:
              <(A, B, C, D, E, F) as diesel::Queryable<(SA, SB, SC, SD, SE, SF), __DB>>
              <(A, B, C, D, E, F) as diesel::Queryable<diesel::sql_types::Record<(SA, SB, SC, SD, SE, SF)>, diesel::pg::Pg>>
    = note: required because of the requirements on the impl of `diesel::Queryable<(diesel::sql_types::Integer, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Integer), diesel::pg::Pg>` for `models::user::User`
    = note: required because of the requirements on the impl of `diesel::query_dsl::LoadQuery<diesel::PgConnection, models::user::User>` for `diesel::query_builder::SelectStatement<schema::users::table, diesel::query_builder::select_clause::DefaultSelectClause, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::WhereClause<diesel::expression::operators::Eq<schema::users::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>>>, diesel::query_builder::order_clause::NoOrderClause, diesel::query_builder::limit_clause::LimitClause<diesel::expression::bound::Bound<diesel::sql_types::BigInt, i64>>>`
```

This is strange...it seems like we obtain the user email from the `User` table and _then_ we update the `email` table...So it seems like we need to figure where it made the changes first.

17.

https://github.com/rust-lang/crates.io/blob/21e04f3028624485621d577359008560728a37d3/src/models/user.rs#L180-L204

```
error[E0026]: struct `models::user::User` does not have a field named `email`
   --> src/models/user.rs:187:13
    |
187 |             email,
    |             ^^^^^ struct `models::user::User` does not have this field

error[E0277]: the trait bound `(i32, std::string::String, std::string::String, std::option::Option<std::string::String>, std::option::Option<std::string::String>, i32): diesel::Queryable<(diesel::sql_types::Integer, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Integer), diesel::pg::Pg>` is not satisfied
   --> src/models/version.rs:118:47
    |
118 |             Some(pb) => users::table.find(pb).first(conn).ok(),
    |                                               ^^^^^ the trait `diesel::Queryable<(diesel::sql_types::Integer, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Integer), diesel::pg::Pg>` is not implemented for `(i32, std::string::String, std::string::String, std::option::Option<std::string::String>, std::option::Option<std::string::String>, i32)`
    |
    = help: the following implementations were found:
              <(A, B, C, D, E, F) as diesel::Queryable<(SA, SB, SC, SD, SE, SF), __DB>>
              <(A, B, C, D, E, F) as diesel::Queryable<diesel::sql_types::Record<(SA, SB, SC, SD, SE, SF)>, diesel::pg::Pg>>
    = note: required because of the requirements on the impl of `diesel::Queryable<(diesel::sql_types::Integer, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Nullable<diesel::sql_types::Text>, diesel::sql_types::Integer), diesel::pg::Pg>` for `models::user::User`
    = note: required because of the requirements on the impl of `diesel::query_dsl::LoadQuery<diesel::PgConnection, models::user::User>` for `diesel::query_builder::SelectStatement<schema::users::table, diesel::query_builder::select_clause::DefaultSelectClause, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::WhereClause<diesel::expression::operators::Eq<schema::users::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>>>, diesel::query_builder::order_clause::NoOrderClause, diesel::query_builder::limit_clause::LimitClause<diesel::expression::bound::Bound<diesel::sql_types::BigInt, i64>>>`
```

Another non-trivial error. I don't think we can just remove `email` from this function entirely because I am pretty sure `EncodablePrivateUser` will need them. So the solutions seems to be that we acquire the email based on the `User`'s `id` instead.

**Comment by hbina** [comment]
If we are ditching `email` in `User`, what about https://github.com/rust-lang/crates.io/blob/3248b4dd9572dc4a9efa4fffe56021cbf59e1199/src/views.rs#L161-L170 ? @carols10cents

**Comment by carols10cents** [comment]
> If we are ditching `email` in `User`, what about this?
> 
> https://github.com/rust-lang/crates.io/blob/3248b4dd9572dc4a9efa4fffe56021cbf59e1199/src/views.rs#L161-L170

We'll need to pass the email into the `encodable_private` method. @hogum has implemented the actual removal of the column from the database, which can be deployed after this PR gets deployed-- parts of [their PR](https://github.com/rust-lang/crates.io/pull/1891) might be relevant here (but the actual dropping of the column shouldn't be moved into this PR)

Let me know if that doesn't answer your questions!

**Comment by carols10cents** [comment]
Hi @hbina, I saw that you had some questions in chat, I'm going to try to answer your questions here :)

This issue is important because we need the Rust code to stop using the `email` column *first*, and we need to deploy that change, so that the servers restart and stop expecting the column to be there before we actually change the database. If we didn't do this, when we do the migration to remove the column, the servers running will have their expectations of the database schema pulled out from under them and that could cause errors.

It's hard to see, but by default, Diesel selects all the columns from a table. For this issue, we need to tell it to select all the columns from the users table except for the email column.

So there will be some temporary code changes needed for this issue that will be undone by #1891, and that's exactly correct. Does that make sense?

I see [this commit](https://github.com/hbina/crates.io/commit/e8b0ad46238c91683bde9416613f4d41defcf9f4) you posted in chat, it's close and I'm going to add some comments on it!

Please ask more questions if I'm still not being clear!

**Comment by hbina** [comment]
Hi! Thanks for the feedback. I have applied the changes that you requested. Is this sufficient? I think I am getting the hang of it. I will try to work on it tonight after school, thank you!

**Comment by thiagoarrais** [comment]
Should this be closed?

**Comment by mugoh** [comment]
This seems to have been solved by #1911 and #1891 if I remember correctly @thiagoarrais

**Comment by JohnTitor** [comment]
Going to close this in favor of the above comments. Thanks for the heads-up!
